### PR TITLE
Improve performance of object-to-object code paths

### LINKED ISSUE #1773: map(object) performance could be improved
http://goo.gl/CIbMkr

I get 0.4 in Chrome ; 0.55 in Firefox.

### PR DESCRIPTION
**Pull request by asaf-romano**
Closes #1773

### CODE REVIEW

**Comment by CrossEye** [comment]
:herb:

File: src/filter.js [hunk 1]
```diff
@@ -36,12 +35,11 @@ var keys = require('./keys');
...
-      _reduce(function(acc, key) {
+      _accumObj(function(key, value, newObject) {
         if (pred(filterable[key])) {
-          acc[key] = filterable[key];
+          newObject[key] = value;
         }
-        return acc;
-      }, {}, keys(filterable)) :
+      }, filterable) :
     // else
```
Comment (jdalton) [inline src/filter.js]: This bypasses bug fixes in `keys`.

Comment (asaf-romano) [inline src/filter.js]: hrm?

Comment (jdalton) [inline src/filter.js]: See https://github.com/ramda/ramda/blob/v0.21.0/src/keys.js#L24-L32.

Comment (asaf-romano) [inline src/filter.js]: The same applies to all those functions that already traverse objects this way (e.g. ``evolve``). I guess copying over these properties doesn't break much. Also note that current implementation of ``keys``, in ES5+ environments, only takes care of the Safari enum bug.

Comment (svozza) [inline src/filter.js]: The hoops Ramda jumps through to support ancient versions of IE seems crazy to me.

Comment (jdalton) [inline src/filter.js]: How many hoops?

Comment (svozza) [inline src/filter.js]: The stuff in `R.keys` and `R.trim` are the ones that spring to mind initially. I've better things to be doing on a Saturday night than trawling through the Ramda source code for more examples. ;)

Comment (jdalton) [inline src/filter.js]: It looks like the bulk of the `R.trim` work is less for old IE and more for consistent whitespace removal.

Comment (svozza) [inline src/filter.js]: Or you could just use the native `trim()` function even though it's not supported in IE8.

Comment (jdalton) [inline src/filter.js]: As I wrote above the bulk of `R.trim` is guarding against buggy native `trim` implementations.

**Comment by asaf-romano** [comment]
Regarding the original issue raised by @jdalton, I'll see what it takes to incorporate the workarounds into the new helper function (without hurting performance). I'm personally more worried about the Safari bug, which seem to apply to the current browser version (I'll have to test, but the code says Safari 9). IE8 is a lost cause.

If it plays out, I may also fix ``evolve`` and other functions to use the new helper. Currently they do not apply any of these workarounds.

**Comment by CrossEye** [comment]
Too long abandoned.  If we need this, we should start afresh.

**Pull request closed without merging**